_posts/2020-03-06-introducing-mirrormaker2.md

```diff
@@ -0,0 +1,268 @@
+
+![Topic renaming with MirrorMaker 2.0](/assets/2020-03-12-mirrormaker-renaming.png)
+
+Why _remote_ topic and not just _source_ topic?
+Because each source cluster can also be a target cluster in a bidirectional configuration.
+This feature also means that it is no longer necessary to aggregate data in a separate cluster.
+
+A rose might be a rose by any other name, but when it comes to replicating topics in MirrorMaker 2.0 we can't afford such ambiguity.
+Particularly if we want to create a bidirectional configuration.
+
+The approach to topic renaming opens up a world of bidirectional opportunities.
+You can now have an _active/active_ cluster configuration that feeds data to and from each cluster.
+For this, you need a MirrorMaker 2.0 cluster at each destination.
+
+### Self-regulating topic replication
+
+Topic replication is defined using regular expression patterns to _whitelist_ or _blacklist_ topics:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1alpha1
+kind: KafkaMirrorMaker2
+metadata:
+  name: my-mirror-maker2
+spec:
+  # ...
+  topicsPattern: ".*"
+```
+
+You use `topicsBlacklistPattern` if you want to use blacklists.
+
+Pretty similar to before, but previously topic _configuration_ was not replicated.
+The target cluster had to be manually maintained to match topic configuration changes in the source cluster.
+
+This time the topic configuration is automatically synchronized between source and target clusters through configuration properties.
+Configuration changes are propagated to remote topics so that new topics and partitions are detected and created.
+The need for rebalancing vanishes.
+
+### Offset tracking and mapping
+
+In the old version of MirrorMaker, the offset of the source topic in the target cluster begins when the replication begins.
+The `__consumer_offsets` topic is not mirrored.
+So offsets of the source topic and its replicated equivalent can have two entirely different positions.
+This can be problematic in a failover situation.
+How to find the offset in the target cluster?
+Strategies such as using timestamps can be adopted, but it adds complexity.
+
+Poof! These issues entirely disappear with MirrorMaker 2.0.
+Instead, we get simplicity and sophistication through the `MirrorCheckpointConnector`.
+
+`MirrorCheckpointConnector` tracks and maps offsets for specified consumer groups using an _offset sync_ topic and _checkpoint_ topic.
+The _offset sync_ topic maps the source and target offsets for replicated topic partitions from record metadata.
+A _checkpoint_ is emitted from each source cluster and replicated in the target cluster through the _checkpoint_ topic.
+The _checkpoint_ topic maps the last committed offset in the source and target cluster for replicated topic partitions in each consumer group.
+
+If you want automatic failover, you can add Kafka's new `RemoteClusterUtils.java` utility class to your consumers.
```

Comment:
Technically, you use the `RemoteClusterUtils` class. But you do not add it by adding the clas to the project. You instead add it by adding the `connect-mirror-client` as dependency. 

```xml
        <dependency>
            <groupId>org.apache.kafka</groupId>
            <artifactId>connect-mirror-client</artifactId>
            <version>2.4.0</version>
        </dependency>
```

Maybe we do not need to show the XML, but I think we need to make this a bit more technically correct.

Reply:
I've changed to:

> If you want automatic failover, you can use Kafka's new `RemoteClusterUtils.java` utility class by adding `connect-mirror-client` as a dependency to your consumers.
> 
> ```xml
> <dependency>
>     <groupId>org.apache.kafka</groupId>
>     <artifactId>connect-mirror-client</artifactId>
>     <version>2.4.0</version>
> </dependency>
> ```